components/util/workqueue.rs

```diff
@@ -190,22 +179,31 @@ impl<QueueData: Sync, WorkData: Send> WorkerThread<QueueData, WorkData> {
     }
 }
 
+/// Erase the user's data type from the WorkerProxy; we don't actually
+/// care what type it is because all we do with the pointer is pass it
+/// back to the user.
+#[inline]
+unsafe fn erase_proxy_type<QueueData>(t: Box<FnBox(&mut WorkerProxy<QueueData>) + Send>)
+                                     -> Box<FnBox(&mut WorkerProxy<c_void>) + Send> {
+    transmute(t)
+}
+
 /// A handle to the work queue that individual work units have.
-pub struct WorkerProxy<'a, QueueData: 'a, WorkData: 'a> {
-    worker: &'a mut Worker<WorkUnit<QueueData, WorkData>>,
+pub struct WorkerProxy<'a, QueueData: 'a> {
+    worker: &'a mut Worker<Box<FnBox(&mut WorkerProxy<c_void>) + Send>>,
```

Comment:
Is this going to result in extra allocations per work unit? This code is designed to not allocate for those.